browse/filters.py

```diff
@@ -0,0 +1,105 @@
+"""Browse jinja filters."""
+import re
+from urllib import parse
+from typing import Match, Callable
+
+from jinja2 import Markup, escape
+from flask import url_for
+
+
+def doi_urls(clickthrough_url_for: Callable[[str], str], text: str) -> str:
+    """Creates links to one or more DOIs.
+
+    clickthrough_url_for is a Callable that takes the URL and returns a clickthrough URL.
+    An example of this is in factory.py
+    While this is called clickthrough_url_for it could be any str -> str fn.
+    For testing this could be the identity function:
+    value = doi_urls( lambda x: x , 'test text bla bla bla')
+    """
+    # How does this ensure escaping?
+    # Two cases:
+    #  1. we get a markup object for text, ex from a previously filter stage
+    #  2. we get a raw str for text, ex meta.abstract
+    # In both of cases the value in result ends up escaped after this conditional.
+    # Then we sub DOI with HTML elements, and return the result as Markup()
+
+    if hasattr(text, '__html__'):
+        result = text
+    else:
+        result = Markup(escape(text))
```

Comment:
Haven't used this yet, but reading the docs, Markup "Marks a string as being safe for inclusion in HTML/XML output without needing to be escaped" - so why do we use the `escape`, then `Markup`? Is this just a way to be more sure that the escaping has indeed happened by specifying it manually?

Reply:
To Jinja, a Markup object is safe to put on the page.  Most operations on a Markup object result in another Markup object and it is smart enough to not double escape. 

When we construct a Markup object, the constructor considers our argument at something we have marked up. We know it is totally safe. So if we get text into our filter that is not a Markup object (that is what that hasattr(text,'__html__') check is about), we need to escape it.

Markup( 'some text <a> stuff' ) will NOT escape the <a> 

Martin had some code that was working with autoescape but I didn't keep it.